crates/core/src/rpc/response.rs

```diff
@@ -1,37 +1,432 @@
-use revision::revisioned;
-use serde::Serialize;
+use std::fmt::Display;
+use std::time::Duration;
 
-use crate::dbs;
-use crate::dbs::Notification;
-use crate::val::{Object, Value};
+use anyhow::Context;
+use revision::{DeserializeRevisioned, Revisioned, SerializeRevisioned, revisioned};
+use serde::{Deserialize, Serialize};
+use surrealdb_types::object;
+use thiserror::Error;
+
+use crate::dbs::QueryResult;
+use crate::rpc::RpcError;
+use crate::types::{
+	PublicArray, PublicKind, PublicNotification, PublicObject, PublicValue, SurrealValue,
+};
+use crate::{dbs, map};
+
+/// Query statistics
+// #[revisioned(revision = 1)]
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[non_exhaustive]
+pub struct DbResultStats {
+	/// The time taken to execute the query.
+	///
+	/// Note: This comes from the `time` field of the [`crate::dbs::Response`] struct.
+	pub execution_time: Option<Duration>,
+}
+
+impl DbResultStats {
+	pub fn with_execution_time(mut self, execution_time: Duration) -> Self {
+		self.execution_time = Some(execution_time);
+		self
+	}
+}
 
 /// The data returned by the database
 // The variants here should be in exactly the same order as `crate::engine::remote::ws::Data`
 // In future, they will possibly be merged to avoid having to keep them in sync.
 #[revisioned(revision = 1)]
-#[derive(Debug, Serialize)]
-pub enum Data {
+#[derive(Debug, Serialize, Deserialize)]
+pub enum DbResult {
 	/// Generally methods return a `expr::Value`
-	Other(Value),
+	Other(PublicValue),
 	/// The query methods, `query` and `query_with` return a `Vec` of responses
-	Query(Vec<dbs::Response>),
+	Query(Vec<dbs::QueryResult>),
 	/// Live queries return a notification
-	Live(Notification),
+	Live(PublicNotification),
 	// Add new variants here
 }
 
-impl Data {
-	pub fn into_value(self) -> Value {
+impl SurrealValue for DbResult {
+	fn kind_of() -> PublicKind {
+		PublicKind::Any
```

Comment:
ideally this type has more detail, although again, I'm not sure if it has a real purpose. We can use the `kind!()` utility though very easy to do with that